ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,654 @@
+#include <Library/UefiDriverEntryPoint.h>
+#include <Protocol/AcpiTable.h>
+#include <Protocol/IoMmu.h>
+#include <Guid/SmmuConfig.h>
+
+#include "IoMmu.h"
+#include "SmmuV3.h"
+#include "SmmuV3Registers.h"
+
+SMMU_INFO  *Smmu;
+
+SMMU_INFO *
+EFIAPI
+SmmuInit (
+  VOID
+  )
+{
+  return (SMMU_INFO *)AllocateZeroPool (sizeof (SMMU_INFO));
+}
+
+VOID
+EFIAPI
+SmmuDeInit (
+  SMMU_INFO  *Smmu
+  )
+{
+  FreePool (Smmu);
+}
+
+STATIC
+VOID
+AcpiPlatformChecksum (
+  IN UINT8  *Buffer,
+  IN UINTN  Size
+  )
+{
+  UINTN  ChecksumOffset;
+
+  ChecksumOffset = OFFSET_OF (EFI_ACPI_DESCRIPTION_HEADER, Checksum);
+
+  // Set checksum field to 0 since it is used as part of the calculation
+  Buffer[ChecksumOffset] = 0;
+
+  Buffer[ChecksumOffset] = CalculateCheckSum8 (Buffer, Size);
+}
+
+/*
+ * A function that add the IORT ACPI table.
+ */
+EFI_STATUS
+AddIortTable (
+  IN EFI_ACPI_TABLE_PROTOCOL  *AcpiTable,
+  IN SMMU_CONFIG              *SmmuConfig
+  )
+{
+  EFI_STATUS            Status;
+  UINTN                 TableHandle;
+  UINT32                TableSize;
+  EFI_PHYSICAL_ADDRESS  PageAddress;
+  UINT8                 *New;
+
+  // Calculate the new table size based on the number of nodes in SMMU_CONFIG struct
+  TableSize = sizeof (SmmuConfig->Config.Iort) +
+              sizeof (SmmuConfig->Config.ItsNode) +
+              sizeof (SmmuConfig->Config.SmmuNode) +
+              sizeof (SmmuConfig->Config.RcNode);
+
+  Status = gBS->AllocatePages (
+                  AllocateAnyPages,
+                  EfiACPIReclaimMemory,
+                  EFI_SIZE_TO_PAGES (TableSize),
+                  &PageAddress
+                  );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to allocate pages for IORT table\n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  New = (UINT8 *)(UINTN)PageAddress;
+  ZeroMem (New, TableSize);
+
+  // Add the ACPI Description table header
+  CopyMem (New, &SmmuConfig->Config.Iort, sizeof (SmmuConfig->Config.Iort));
+  ((EFI_ACPI_DESCRIPTION_HEADER *)New)->Length = TableSize;
+  New                                         += sizeof (SmmuConfig->Config.Iort);
+
+  // ITS Node
+  CopyMem (New, &SmmuConfig->Config.ItsNode, sizeof (SmmuConfig->Config.ItsNode));
+  New += sizeof (SmmuConfig->Config.ItsNode);
+
+  // SMMUv3 Node
+  CopyMem (New, &SmmuConfig->Config.SmmuNode, sizeof (SmmuConfig->Config.SmmuNode));
+  New += sizeof (SmmuConfig->Config.SmmuNode);
+
+  // RC Node
+  CopyMem (New, &SmmuConfig->Config.RcNode, sizeof (SmmuConfig->Config.RcNode));
+  New += sizeof (SmmuConfig->Config.RcNode);
+
+  AcpiPlatformChecksum ((UINT8 *)(UINTN)PageAddress, TableSize);
+
+  Status = AcpiTable->InstallAcpiTable (
+                        AcpiTable,
+                        (EFI_ACPI_COMMON_HEADER *)(UINTN)PageAddress,
+                        TableSize,
+                        &TableHandle
+                        );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to install IORT table\n"));

```

Comment:
should you free the memory allocations?

Reply:
I am now freeing everything inside SmmuDeInit() and that is called after if any of the functions fail inside the Smmu driver entry point.